grudge/trace_pair.py

```diff
@@ -272,61 +274,122 @@ def connected_ranks(dcoll: DiscretizationCollection):
     return get_connected_partitions(dcoll._volume_discr.mesh)
 
 
+def flatten_to_numpy_from_container(actx, ary):
+    """
+    """
+    if is_array_container(ary):
+        containerized_data_template = {}
+        flat_data = []
+        for key, value in serialize_container(ary):
+            # FIXME: Need to special case nested object arrays
+            if isinstance(value, np.ndarray) and value.dtype == "O":
+                numpy_values = flatten_to_numpy(actx, value)
+                flat_data.append(np.concatenate(numpy_values))
+            else:
+                numpy_values = actx.to_numpy(value)
+                flat_data.append(numpy_values)
+            containerized_data_template[key] = numpy_values
+        return np.concatenate(flat_data), containerized_data_template
+    else:
+        # No data template for arys with only a single state
+        # e.g. a single DOFArray
+        return flatten_to_numpy(actx, ary), None
+
+
+def unflatten_from_numpy_to_container(
+        actx, numpy_ary, ary_container_data, discr, template):
+    """
+    """
+    if not isinstance(template, DOFArray):
+        remote_container_data = {}
+        offset = 0
+        for key, _ in serialize_container(template):
+            _value = ary_container_data[key]
+            # FIXME: Need to inspect value before it was flattened to determine
+            # the right offsets
+            if isinstance(_value, np.ndarray) and _value.dtype == "O":
+                remote_comps = []
+                for _subary_idx in range(len(_value)):
+                    _subary = _value[_subary_idx]
+                    _subary_ndofs, = _subary.shape
+                    remote_comps.append(
+                        unflatten_from_numpy(
+                            actx, discr,
+                            numpy_ary[offset:offset+_subary_ndofs]
+                        )
+                    )
+                    offset += _subary_ndofs
+                remote_data = make_obj_array(remote_comps)
+            else:
+                _ndofs, = _value.shape
+                remote_data = unflatten_from_numpy(
+                    actx, discr, numpy_ary[offset:offset+_ndofs]
+                )
+                offset += _ndofs
+            remote_container_data[key] = remote_data
+        return deserialize_container(
+            template,
+            remote_container_data.items()
+        )
+    else:
+        return unflatten_from_numpy(actx, discr, numpy_ary)
```

Comment:
@inducer @alexfikl This is my ugly attempt and writing an inverse transformation of `flatten_to_numpy_from_container`. I think this function probably needs the most scrutiny, as there may be better ways of doing this that I haven't quite realized yet. Or this may reveal something that's missing in `meshmode.dof_array` / `arraycontext.traversal`?

Reply:
Note that `ary_container_data` in this function is what was previously returned as `containerized_data_template` in `flatten_to_numpy_from_container`. I should probably use consistent naming.

Reply:
This commit https://github.com/inducer/grudge/pull/154/commits/9127937adb3480204192813fa5de3f80be99109a cleans things up